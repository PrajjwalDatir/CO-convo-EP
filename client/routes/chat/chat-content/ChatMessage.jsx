import React from 'react';
import PropTypes from 'prop-types';
import { DateTime } from 'luxon';
import * as R from 'ramda';
import useEffectOnce from '#/hooks/useEffectOnce';

function ChatMessage({ message }) {
  const isSafeToUpdate = React.useRef(true);
  const [whenMessageWasSent, setWhenMessageWasSent] = React.useState(null);
  const [shouldRecalculateTime, setShouldRecalculateTime] = React.useState(
    true,
  );

  const recalculateTimeInAMinute = React.useCallback(() => {
    setTimeout(() => {
      if (isSafeToUpdate.current) {
        setShouldRecalculateTime(true);
      }
    }, 60 * 1000);
  }, [setShouldRecalculateTime]);

  React.useEffect(() => {
    if (shouldRecalculateTime) {
      setShouldRecalculateTime(false);

      const sent = DateTime.fromISO(message.sent);

      const minutesSinceSend = DateTime.local()
        .diff(sent, 'minutes')
        .toObject().minutes;

      R.o(
        setWhenMessageWasSent,
        R.cond([
          [R.gte(1), R.o(() => 'Just now', recalculateTimeInAMinute)],
          [R.gte(10), R.o(() => sent.toRelative(), recalculateTimeInAMinute)],
          [R.T, () => sent.toLocaleString(DateTime.DATETIME_MED)],
        ]),
      )(minutesSinceSend);
    }
  }, [
    message.sent,
    whenMessageWasSent,
    shouldRecalculateTime,
    recalculateTimeInAMinute,
  ]);

  useEffectOnce(() => () => {
    isSafeToUpdate.current = false;
  });

  return (
    <div className="my-3">
      <div>
        <span className="font-bold mr-1">
          {message.user ? message.user.username : <>&nbsp;</>}
        </span>
        <span className="text-xs text-gray-700">
          <time dateTime={message.sent}>{whenMessageWasSent}</time>
        </span>
      </div>
      <div className="whitespace-pre-line">{message.message}</div>
    </div>
  );
}

ChatMessage.propTypes = {
  message: PropTypes.shape({
    user: PropTypes.shape({
      username: PropTypes.string,
    }),
    sent: PropTypes.string, // ISO date
    message: PropTypes.string,
  }).isRequired,
};

export default React.memo(ChatMessage);
