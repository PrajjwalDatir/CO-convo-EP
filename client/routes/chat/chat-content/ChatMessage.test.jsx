import React from 'react';
import { mount } from 'enzyme';
import ChatMessage from './ChatMessage';

describe('<ChatMessage />', () => {
  it('should match snapshot', () => {
    const wrapper = mount(
      <ChatMessage
        message={{
          user: { username: 'user' },
          sent: '2019-11-23T21:00:39.290Z',
          message: 'hi',
        }}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
