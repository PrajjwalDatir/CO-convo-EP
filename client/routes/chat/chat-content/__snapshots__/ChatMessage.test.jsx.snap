// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMessage /> should match snapshot 1`] = `
<Memo(ChatMessage)
  message={
    Object {
      "message": "hi",
      "sent": "2019-11-23T21:00:39.290Z",
      "user": Object {
        "username": "user",
      },
    }
  }
>
  <div
    className="my-3"
  >
    <div>
      <span
        className="font-bold mr-1"
      >
        user
      </span>
      <span
        className="text-xs text-gray-700"
      >
        <time
          dateTime="2019-11-23T21:00:39.290Z"
        >
          24 Nov 2019, 2:30 am
        </time>
      </span>
    </div>
    <div
      className="whitespace-pre-line"
    >
      hi
    </div>
  </div>
</Memo(ChatMessage)>
`;
